File: LC/lc977/lc977.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortedSquares = function(nums) {
  /* let a = nums.map(x => x*x), p, tmp1, tmp2, res = [];
  for (i = 0; i < nums.length-1; i++) {
      if (a[i] < a[i+1]) {
          p = i;
          break;
      }
  }
  let b = a.splice(p+1);
  b.reverse();
  tmp1 = a.pop(), tmp2 = b.pop();
  while (a.length || b.length){
      if (tmp1 && tmp2) {
          if (tmp1 < tmp2){
              res.push(tmp1);
              tmp1 = a.pop();
          } else {
              res.push(tmp2);
              tmp2 = b.pop();
          } 
      }
      else if (tmp1) {
          res.push(tmp1);
          tmp1 = a.pop();
      } else {
          res.push(tmp2);
          tmp2 = b.pop();
      }
  }
  return res; */
  return nums.map(x => x*x).sort((a, b) => a- b);
};

File: LC/lc977/lc977.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        def merge(a, b):
            L1, L2 = len(a), len(b)
            p1, p2 = 0, 0
            res = list()
            while p1 < L1 and p2 < L2:
                if a[p1] < b[p2]:
                    res.append(a[p1])
                    p1 += 1
                else:
                    res.append(b[p2])
                    p2 += 1
            if p1 < L1:
                res.extend(a[p1:])
            if p2 < L2:
                res.extend(b[p2:])
            return res
        p = 0
        def square(x): return x*x
        while p < len(nums) and nums[p]<0:
            p+= 1
        return merge(list(map(square, reversed(nums[:p]))), list(map(square,nums[p:])))
             